HPCC-19923 Implement copy/paste support in Config Manager 2.0

kenrowland opened the pull request:
Add serialize method to serialize environment subtree
Add method to insert serialized subtree into an environment
Add ESP service methods to copy a node and insert a node copy

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

kenrowland commented:
@jpmcmu Please review. Thanks.

HPCCSmoketest commented:
This PR (12105) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12105, label: HPCC-19923 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 78234b1aea62905fc445790cf91254e16716dbce
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-12105/HPCC-Platform/esp/services/ws_configmgr/ws_configmgrService.cpp:26:58: fatal error: ../../../../release/generated/ws_configmgr.hpp: No such file or directory
make[2]: *** [esp/services/ws_configmgr/CMakeFiles/ws_configmgr.dir/ws_configmgrService.cpp.o] Error 1

[Build error:](https://gist.github.com/719914b861f2b2de153df124466b6d18)
Build log [Build log](https://gist.github.com/HPCCSmoketest/719914b861f2b2de153df124466b6d18#file-result-19-01-28-14-47-11-log)

kenrowland commented:
@AttilaVamos I know not why this would fail. The file in question is a generated file from the ecm file. This builds w/o error in my dev environment. Unless there is additional logging that I cannot see, could this be something else causing it?

HPCCSmoketest commented:
Process of PR-12105, label: HPCC-19923 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: dc59814226145ea5ec6a67cccebd6635afed0c15
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-12105/HPCC-Platform/esp/services/ws_configmgr/ws_configmgrService.cpp:26:58: fatal error: ../../../../release/generated/ws_configmgr.hpp: No such file or directory
make[2]: *** [esp/services/ws_configmgr/CMakeFiles/ws_configmgr.dir/ws_configmgrService.cpp.o] Error 1

[Build error:](https://gist.github.com/bb47425a5c640ee11a728a6428b92013)
Build log [Build log](https://gist.github.com/HPCCSmoketest/bb47425a5c640ee11a728a6428b92013#file-result-19-01-29-11-11-40-log)

AttilaVamos reviewed (commented):
In  esp/services/ws_configmgr/ws_configmgrService.cpp there is an include:

```
#include "../../../../release/generated/ws_configmgr.hpp"
```
I don't think this path is correct because the ws_configmgr directory is in the third level down from the source root and this path references to the 4th level up. If it is then (I think) it means you already have a "release" directory above the source root ("HPCC-Platform") with generated header file.

The other way, why I think it is not correct because the build directory can be anywhere (same or different level from the source "HPCC-Platform" directory) and can called by any name e.g. "build" or "release" and from that build directory can have a "generated" directory.

(And you have an unresolved conflict in this commit)

Commit:
HPCC-19923 Implement copy/paste support in Config Manager 2.0

Add serialize method to serialize environment subtree
Add method to insert serialized subtree into an environment
Add ESP methods for copy and paste

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

kenrowland commented:
@AttilaVamos Yes I saw that. Not sure how that line got in there. Perhaps the tools I use thought it needed to be there. I was especially odd because it was referencing a release build directory directly. I have removed it.

HPCCSmoketest commented:
Process of PR-12105, label: HPCC-19923 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 56804c8f467e61ce07728f9c449fe32e545d42dd
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |813 |813 |0 |
|test (thor) |739 |739 |0 |
|test (roxie) |887 |887 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  195 sec (00:03:15) |    0 sec (00:00:00) |    2 sec (00:00:02) |   16 sec (00:00:16) | 1462 sec (00:24:22) |   19 sec (00:00:19) | 1709 sec (00:28:29) |

AttilaVamos commented:
@kenrowland back to you.

kenrowland commented:
@richardkchapman Please merge

richardkchapman commented:
Has this been reviewed? @jpmcmu ?

kenrowland commented:
@richardkchapman sorry confused with another PR. James will review.

jpmcmu commented on configuration/configmgr/configmgrlib/EnvironmentNode.cpp:
```diff
@@ -217,66 +217,53 @@ std::string EnvironmentNode::getLocalValue() const
-            // If this value must be unique, make sure it is
-            if (attrIt->second->getSchemaValue()->isUniqueValue())
+            bool found = false;
+            std::vector<std::string> allValues;
```
Not required but allValues an uniqueValues could be hoisted outside of the for loop and cleared to avoid memory churn a bit.

jpmcmu commented on configuration/configmgr/configmgrlib/SchemaItem.cpp:
```diff
@@ -646,3 +647,93 @@ void SchemaItem::setRequiredInstanceComponents(const std::string list)
+    // maintain a parent child relationship like the environment, the environment nodes for this schema item need to
+    // be separated by each environment node's parent. Then the number of nodes, by parent, is compared agains the
+    // allowed min/max range
+    std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> envNodesByParent;
```
It looks like this is used to track the number of children for a given parent. Could this become ``std::map<std::string,uint32_t>`` to avoid having a std::vector in the map? Granted this is probably less of a concern as vector should have move semantics

kenrowland replied:
The observation is a good one. When writing the code, there was the potential need to inspect each of the nodes by type. While that has not materialized, I would prefer to leave the construct the way it is in case that need arises again.

jpmcmu commented on esp/services/ws_configmgr/ws_configmgrService.cpp:
```diff
@@ -352,7 +355,7 @@ bool Cws_configMgrEx::onGetNode(IEspContext &context, IEspGetNodeRequest &req, I
 
     getNodeResponse(pNode, resp);
-    pNode->validate(status, false);  // validate this node only
+    pNode->getSchemaItem()->validate(status, false);  // validate this node only
```
Can schemaItem be a nullptr?

kenrowland replied:
No. I ensure that all nodes have a schema item. This is even the case if no schema is defined. A default schema item is constructed for this case.

jpmcmu commented on esp/services/ws_configmgr/ws_configmgrService.cpp:
```diff
@@ -593,6 +596,103 @@ bool Cws_configMgrEx::onFetchNodes(IEspContext &context, IEspFetchNodesRequest &
+    std::string sessionId = req.getSessionId();
+    Status status;
+
+    ConfigMgrSession *pSession = getConfigSession(sessionId, true);
```
Do we need a nullptr check on pSession? It looks like sessionId is coming from the user

kenrowland replied:
No. The getConfigSession method will throw an exception if the ID is not valid.

jpmcmu commented on esp/services/ws_configmgr/ws_configmgrService.cpp:
```diff
@@ -593,6 +596,103 @@ bool Cws_configMgrEx::onFetchNodes(IEspContext &context, IEspFetchNodesRequest &
+    ConfigMgrSession *pSession = getConfigSession(sessionId, true);
+
+    if (nodeId.empty())
+        nodeId = pSession->m_pEnvMgr->getRootNodeId();
```
Can m_pEnvMgr be nullptr?

kenrowland replied:
No, provided 'true' was passed as the second parameter to getConfigSession. That indicates that a loaded environment is required.

jpmcmu reviewed (commented):
@kenrowland a couple of comments / questions

kenrowland reviewed (commented):
@jpmcmu Please see my comments. If you are OK with the responses, please approve. Thanks for the review.

jpmcmu commented:
@kenrowland looks good

kenrowland commented:
@richardkchapman please merge. Thanks.
